lib/tables.c

```diff
@@ -2141,7 +2141,13 @@ individual_table_dump_text(individual_table_t *self, FILE *out)
             goto out;
         }
         for (k = self->location_offset[j]; k < self->location_offset[j + 1]; k++) {
-            err = fprintf(out, "%f", self->location[k]);
+            // necessary precision for non-lossiness;
+            // see https://stackoverflow.com/a/19897395/2752221
+#ifdef DBL_DECIMAL_DIG
```

Comment:
> It's impossible to print out in decimal form losslessly anyway

Not sure what you mean by this; this is precisely what the use of DBL_DECIMAL_DIG achieves.

It's important for SLiM testing to be able to round-trip out to text files and back in again (the ability to do that has already allowed me to catch one important bug that would have been very difficult to debug using binary output files).

As to the conditional compilation, agreed, it is ugly.  In the Eidos headers I have this:

// The question here is how many digits are needed to print out in text representations of floating-point numbers
// in order to preserve their exact value.  This is quite a complicated question, and the macros in float.h needed
// to resolve it are not always available.  See the answer from chux at https://stackoverflow.com/a/19897395/2752221,
// which this is based upon.  The EIDOS_DBL_DIGS and EIDOS_FLT_DIGS macros defined here can be used directly with
// a printf() format of %.*g with a double or float argument, respectively, to produce a correct representation in
// all cases, if I understand that thread correctly.

#ifdef DBL_DECIMAL_DIG
#define EIDOS_DBL_DIGS (DBL_DECIMAL_DIG)
#else  
#define EIDOS_DBL_DIGS (DBL_DIG + 3)
#endif

#ifdef FLT_DECIMAL_DIG
#define EIDOS_FLT_DIGS (FLT_DECIMAL_DIG)
#else  
#define EIDOS_FLT_DIGS (FLT_DIG + 3)
#endif

And then in the rest of the code in Eidos and SLiM I can just use EIDOS_DBL_DIGS or EIDOS_FLT_DIGS, as the case may be.  If you don't want the conditional compilation in the function's code, perhaps you could adopt this sort of scheme to hide the issue; I would have done that myself in these diffs, but I thought you would prefer the present diffs since the issue is only at one spot in the code.  If you introduce such a header-defined constant, then at that point it shouldn't introduce any sort of testing issue for you, since the only thing that would be affected would be the number of digits emitted for such floating-point numbers; since you not concerned with losslessness for those anyway, I imagine that ought to be a non-issue for your testing...?

Finally, re: "I don't want to have to go through every other place we print out double values", this is the only place in the code that SLiM cares about; I already checked.  Most things that msprime considers to be floating-point (times, chromosome positions) are integers in SLiM, so round-tripping them is lossless as presently implemented, for SLiM's usage case.

But if you do want to just use a constant precision specifier, then please use "%.17g", not "%.14g"; that is necessary for lossless round-trips, as explained on the linked stackoverflow thread.  :->  If round-trips are lossy then various tests on my side fail and it becomes impossible to use round-tripping for testing.  Thanks!

Reply:
(But of course if the number of bits used by float/double ever increase, "%.17g" would no longer do the right thing, so to my mind that is not a good solution...)

Reply:
I'm fine with %17g --- the size of a double is driven by IEEE semantics and this is never going to change.  We can add a macro of the number of digits (i.e. ``printf("x = %.g" MSP_DBL_DECIMAL_DIG, x)``). I don't want to pepper the code with lots of conditional compiles though.

> It's impossible to print out in decimal form losslessly anyway

I was just referring to the fact that decimal fractions like 1/10 can't be expressed in a finite number of bits in a binary representation. I had thought that the conversion from decimal -> binary -> decimal is inherently lossy. But it turns out that %.17g does the trick --- you learn something new every day!

Reply:
OK, sounds good.  @petrelharp let me know if you want me to do something; I'm not sure how to contribute changes to a pull request.  But it sounds like Jerome and I are in agreement that something like this could go into a header:
```
#ifdef DBL_DECIMAL_DIG
#define MSP_DBL_DECIMAL_DIG (DBL_DECIMAL_DIG)
#else
#define MSP_DBL_DECIMAL_DIG (DBL_DIG + 3)
#endif
```
and then this stuff in your pull:
```
            // necessary precision for non-lossiness; see https://stackoverflow.com/a/19897395/2752221
#ifdef DBL_DECIMAL_DIG
            err = fprintf(out, "%.*g", DBL_DECIMAL_DIG, self->location[k]);
#else
            fprintf(out, "%.*g", (DBL_DIG + 3), self->location[k]);
#endif
```
would just change to:
```
err = fprintf(out, "%.*g", MSP_DBL_DECIMAL_DIG, self->location[k]);
```
Sound good?

Reply:
sounds good - could you make the changes to the files on the SLiM side, @bhaller, and I'll move 'em over?

Reply:
Done, just committed.

Reply:
I've just merged those over.